channeldb/invoices.go

```diff
@@ -439,51 +449,57 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
 		InvoiceQuery: q,
 	}
 
-	// If the caller provided an index offset, then we'll not know how many
-	// records we need to skip. We'll also keep track of the record offset
-	// as that's part of the final return value.
-	invoicesToSkip := q.IndexOffset
-	invoiceOffset := q.IndexOffset
-
 	err := d.View(func(tx *bolt.Tx) error {
 		// If the bucket wasn't found, then there aren't any invoices
 		// within the database yet, so we can simply exit.
 		invoices := tx.Bucket(invoiceBucket)
 		if invoices == nil {
 			return ErrNoInvoicesCreated
 		}
-		invoiceAddedIndex := invoices.Bucket(addIndexBucket)
-		if invoiceAddedIndex == nil {
+		invoiceAddIndex := invoices.Bucket(addIndexBucket)
+		if invoiceAddIndex == nil {
 			return ErrNoInvoicesCreated
 		}
 
-		// We'll be using a cursor to seek into the database, so we'll
-		// populate byte slices that represent the start of the key
-		// space we're interested in.
-		var startIndex [8]byte
-		switch q.PendingOnly {
-		case true:
-			// We have to start from the beginning so we know
-			// how many pending invoices we're skipping.
-			byteOrder.PutUint64(startIndex[:], uint64(1))
-		default:
-			// We can seek right to the invoice offset we want
-			// to start with.
-			invoicesToSkip = 0
-			byteOrder.PutUint64(startIndex[:], uint64(invoiceOffset+1))
+		// keyForIndex is a helper closure that retrieves the invoice
+		// key for the given add index of an invoice.
+		keyForIndex := func(c *bolt.Cursor, index uint64) []byte {
+			var keyIndex [8]byte
+			byteOrder.PutUint64(keyIndex[:], index)
+			_, invoiceKey := c.Seek(keyIndex[:])
+			return invoiceKey
+		}
+
+		// nextKey is a helper closure to determine what the next
+		// invoice key is when iterating over the invoice add index.
+		nextKey := func(c *bolt.Cursor) ([]byte, []byte) {
+			if q.Reversed {
+				return c.Prev()
+			}
+			return c.Next()
+		}
+
+		// We'll be using a cursor to seek into the database and return
+		// a slice of invoices. We'll need to determine where to start
+		// our cursor depending on the parameters set within the query.
+		c := invoiceAddIndex.Cursor()
+		invoiceKey := keyForIndex(c, q.IndexOffset+1)
```

Comment:
Does this mean the invoiceKeys are 1-indexed? Don't see how you can fetch invoice `0` now.

Reply:
The add_index value starts at 1.

Reply:
Yep, the invoice add index starts at 1, so it's safe to do this.